### ISSUE TITLE
Flag leak

### ISSUE THREAD

**Issue by royragsdale** [picoCTF/picoCTF#43]
Currently the `/api/problems/<pid>` route can be used to leak flags:

This is because the `get_single_problem_hook` directly returns the full results from a database query.
[code](https://github.com/picoCTF/picoCTF/blob/master/picoCTF-web/api/routes/problem.py#L66) 

A user can determine the problem pid by viewing the requests made when navigating to the problems tab.
Example leak

```
$ curl 'http://192.168.2.2/api/problems/1bef644c399e10a3f35fecdbf590bd0c'  -H 'Cookie: flask=.eJwNyjkKgEAMAMC_pLbIrjFZ_E1OEEEbrcS_azPVPHCdex6wwoIydG6Vjk6uqIQm2VkRLVQYJri3-KMUR1Ef44fMfEQ1LeGe5UZc8H4fsRih.CehO3A.aueQhjHWsvJyp9b_-uxyIWyK_Gg'
{"status": 1, "data": {"hints": ["This is a classic buffer overflow with no modern protections."], "name": "Buffer Overflow 1", "organization": "ForAllSecure", "author": "Tim Becker", "instances": [{"description": "Exploit the <a href='//192.168.2.3/static/c95410042007bb17f49b891a2a87afb2/vuln'>Buffer Overflow</a> found here: /problems/buffer-overflow-1_1.", "flag_sha1": "5675d2d5819084d4203c1ef314239527074938a9", "instance_number": 1, "sid": "c5c8c58c69636756d94899c387cd99e0", "user": "buffer-overflow-1_1", "files": [{"path": "flag.txt", "user": null, "group": null, "permissions": 288}, {"path": "vuln", "user": null, "group": null, "permissions": 1517}], "flag": "35013564b97b80d4fd3f2be45e5836ff", "iid": "691b52dd3fa57d7a93e2941706462608", "server": "192.168.2.3", "should_symlink": true, "socket": null, "deployment_directory": "/problems/buffer-overflow-1_1", "service": "buffer-overflow-1_1.service"}, {"description": "Exploit the <a href='//192.168.2.3/static/bd08ee41f495f8bff378c13157d0f511/vuln'>Buffer Overflow</a> found here: /problems/buffer-overflow-1_0.", "flag_sha1": "4b97abef055a11ec19c14622eb31eb1168d98aca", "instance_number": 0, "sid": "c5c8c58c69636756d94899c387cd99e0", "user": "buffer-overflow-1_0", "files": [{"path": "flag.txt", "user": null, "group": null, "permissions": 288}, {"path": "vuln", "user": null, "group": null, "permissions": 1517}], "flag": "638608c79eca2165e7b241ff365df05b", "iid": "c1916a37d2677697345c04682170d438", "server": "192.168.2.3", "should_symlink": true, "socket": null, "deployment_directory": "/problems/buffer-overflow-1_0", "service": "buffer-overflow-1_0.service"}], "pid": "1bef644c399e10a3f35fecdbf590bd0c", "sanitized_name": "buffer-overflow-1", "category": "Binary Exploitation", "description": "Exploit the {{url_for(\"vuln\", display=\"Buffer Overflow\")}} found here: {{directory}}.", "disabled": false, "score": 50}, "message": null
```

**Comment by carlislemc** [comment]
This one looks pretty urgent. I just confirmed it is still an issue. To reproduce:
1) login and use something something like burpsuite to see requests/replies
2) Track a request like: GET /api/problems?_=1531353833777 
3) Look in the reply to get the PID: "pid": "badaea98b298f181c90ad20a02259222"
4) Get the cookie (which now has flask and token) and do a curl as above.

**Comment by carlislemc** [comment]
This may be as simply as doing a require_admin near line 82 of:

https://github.com/picoCTF/picoCTF/blob/master/picoCTF-web/api/routes/problem.py

**Comment by hiliang-cmu** [comment]
Fixed by #120 but

> At some point we should revisit for consistency and what should/needs to be exposed at these endpoints, but for now this works both for sensitivity of data and for functionality of the (2018) game.
